precompiles/proxy/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+	<<Runtime as frame_system::Config>::Call as Dispatchable>::Origin:
+		From<Option<Runtime::AccountId>>,
+	<Runtime as frame_system::Config>::Call: From<ProxyCall<Runtime>>,
+{
+	fn execute(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let selector = handle.read_selector()?;
+
+		handle.check_function_modifier(match selector {
+			Action::IsProxy => FunctionModifier::View,
+			_ => FunctionModifier::NonPayable,
+		})?;
+
+		match selector {
+			Action::AddProxy => Self::add_proxy(handle),
+			Action::RemoveProxy => Self::remove_proxy(handle),
+			Action::RemoveProxies => Self::remove_proxies(handle),
+			Action::IsProxy => Self::is_proxy(handle),
+		}
+	}
+}
+
+impl<Runtime> ProxyWrapper<Runtime>
+where
+	Runtime: pallet_proxy::Config + pallet_evm::Config + frame_system::Config,
+	<<Runtime as pallet_proxy::Config>::Call as Dispatchable>::Origin:
+		From<Option<Runtime::AccountId>>,
+	<Runtime as pallet_proxy::Config>::ProxyType: Decode,
+	<Runtime as frame_system::Config>::Call:
+		Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<<Runtime as frame_system::Config>::Call as Dispatchable>::Origin:
+		From<Option<Runtime::AccountId>>,
+	<Runtime as frame_system::Config>::Call: From<ProxyCall<Runtime>>,
+{
+	/// Register a proxy account for the sender that is able to make calls on its behalf.
+	/// The dispatch origin for this call must be Signed.
+	///
+	/// Parameters:
+	/// * delegate: The account that the caller would like to make a proxy.
+	/// * proxy_type: The permissions allowed for this proxy account.
+	/// * delay: The announcement period required of the initial proxy. Will generally be zero.
+	fn add_proxy(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		input.expect_arguments(3)?;
+
+		let delegate = input
+			.read::<Address>()
+			.map(|addr| Runtime::AddressMapping::into_account_id(addr.into()))?;
+		let proxy_type = input
+			.read::<u8>()
+			.map_err(|_| revert("failed reading proxy_type"))
+			.and_then(|value| {
+				Runtime::ProxyType::decode(&mut value.to_le_bytes().as_slice())
+					.map_err(|_| revert("failed decoding proxy_type"))
+			})?;
+		let delay = input.read::<u32>()?.into();
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+
+		// Disallow re-adding proxy via precompile to prevent privilege escalation attack.
+		// If a proxy account manages to call this precompile via `pallet_proxy.proxy` they can
+		// escalate their own privileges.
+		// See: https://github.com/PureStake/sr-moonbeam/issues/30
+		// Note: It is also assumed that EVM calls are only allowed through `Origin::Root` and
+		// filtered via CallFilter
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		if ProxyPallet::<Runtime>::proxies(&origin)
+			.0
+			.iter()
+			.any(|pd| pd.delegate == delegate)
+		{
+			return Err(revert("Duplicate"));
```

Comment:
A more precise error message would be welcome

Reply:
It's the same error the proxy pallet would've [returned](https://docs.rs/pallet-proxy/latest/src/pallet_proxy/lib.rs.html#576) - should I still make it verbose?

Reply:
No it's not the same, because the proxy pallet only returns this error if there is already a proxy of the **same type with the same delay**. Here the error is returned in many more cases, it is not a duplication problem